query-parser/src/parse.rs

```diff
@@ -93,34 +96,75 @@ impl<I> Query<I>
 
 def_value_satisfy_parser_fn!(Query, variable, Variable, Variable::from_value);
 def_value_satisfy_parser_fn!(Query, source_var, SrcVar, SrcVar::from_value);
+def_value_satisfy_parser_fn!(Query, predicate_fn, PredicateFn, PredicateFn::from_value);
+def_value_satisfy_parser_fn!(Query, fn_arg, FnArg, FnArg::from_value);
 
 pub struct Where<I>(::std::marker::PhantomData<fn(I) -> I>);
 
-def_value_satisfy_parser_fn!(Where, pattern_value_place, PatternValuePlace, PatternValuePlace::from_value);
-def_value_satisfy_parser_fn!(Where, pattern_non_value_place, PatternNonValuePlace, PatternNonValuePlace::from_value);
+def_value_satisfy_parser_fn!(Where,
+                             pattern_value_place,
+                             PatternValuePlace,
+                             PatternValuePlace::from_value);
+def_value_satisfy_parser_fn!(Where,
+                             pattern_non_value_place,
+                             PatternNonValuePlace,
+                             PatternNonValuePlace::from_value);
+
+/// Take a vector Value containing one vector Value, and return the `Vec` inside the inner vector.
+/// Also accepts an inner list, returning it as a `Vec`.
+fn unwrap_nested(x: edn::Value) -> Option<Vec<edn::Value>> {
+    match x {
+        edn::Value::Vector(mut v) => {
+            match v.pop() {
+                Some(edn::Value::List(items)) => Some(items.into_iter().collect()),
+                Some(edn::Value::Vector(items)) => Some(items),
+                _ => None,
+            }
+        }
+        _ => None,
+    }
+}
+
+/// A vector containing just a parenthesized filter expression.
+def_value_parser_fn!(Where, pred, WhereClause, input, {
+    satisfy_map(|x: edn::Value| {
+        // Accept either a list or a vector here:
+        // `[(foo ?x ?y)]` or `[[foo ?x ?y]]`
```

Comment:
Indeed, the grammar declares that anywhere you see `[]` it means a list or vector (probably because all Clojure cares about are `seq`s). The common usages we see of things like `[(< ?x 5)]` are just convention.

I don't think we should jump to being so flexible; I'm happy to pave the cowpaths and make `[]` required in most cases.